feat(schematics): add entity generation as part of feature schematic

wgd3 opened the pull request:
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngrx/platform/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Documentation has been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

If a user wishes to create an entity as part of a new feature state, they must run multiple commands:

```sh
ng generate feature User

ng generate entity User -m app.module.ts -r user.reducer.ts
```

Closes #3260 

## What is the new behavior?

Users can now optionally generate an entity when running the `feature` schematic:

```sh
ng generate feature User --entity
```

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information

This functionality does _not_ allow for custom entity names - instead, a boolean is passed to the `feature` schematic. If this needs to be changed I'm happy to update this PR allowing for custom entity names.

timdeschryver commented on modules/schematics/src/feature/schema.json:
```diff
@@ -65,6 +65,13 @@
+      "description": "Toggle whether an entity is created as part of the feature",
+      "type": "boolean",
+      "aliases": ["e"],
+      "x-prompt": "Specifies whether an entity is created as part of the feature",
```
To streamline this with the other prompts we should reword this into a question:

```suggestion
      "x-prompt": "Should we use @ngrx/entity to create the reducer?",
```

wgd3 commented on modules/schematics/src/selector/schema.json:
```diff
@@ -5,7 +5,7 @@
   "properties": {
     "name": {
-      "description": "The name of the selector.",
+      "description": "The name of the f-ing selector.",
```
Well that's extremely embarrassing - apologies for overlooking this. Some late night frustrations resulted testing various strings.

wgd3 commented on modules/schematics/src/feature/index.ts:
```diff
@@ -9,51 +9,69 @@ import { Schema as FeatureOptions } from './schema';
+          api: options.api,
+          prefix: options.prefix,
+        }),
+        schematic('reducer', {
```
That makes sense! I may have been focused on the small stuff rather than the overall flow of the generator. I resorted to using `MergeStrategy.Overwrite` to try and avoid duplicated actions/reducers, which in retrospect, is a relatively aggressive way of handling code generation. 

I'll try this approach and update the PR.

timdeschryver replied:
Instead of overwriting could we also skip the reducer and action schematics when using the `entity` flag?

wgd3 commented on modules/schematics/src/reducer/index.ts:
```diff
@@ -60,9 +60,22 @@ export default function (options: ReducerOptions): Rule {
+        chain([
+          addReducerToState({
+            ...options,
+            plural: options.feature && options.entity,
```
This was related to the comment below - a way to conditionally handle plurals when including entity generation. It will be reverted with the next update.

wgd3 commented:
@timdeschryver I've refactored my local branch based on your recommendations - that turned out to be a much simpler change to the code base than my initial commits. Before I push my changes, I wanted to get your workflow preference:

There were enough changes between my commits and the v16 release (congrats!) that I ran `git reset --hard upstream/main` to start fresh and up-to-date. Then applied my changes on there. Are you/the team ok with a force-push to this branch? Or would you prefer I close this PR and open a new one?

timdeschryver commented:
Thanks in advance @wgd3.
I'm currently a bit busy but I'll review it later this week.

timdeschryver commented on yarn.lock:
```diff
@@ -2,33 +2,25 @@
 
 
-"@ampproject/remapping@2.2.1":
+"@ampproject/remapping@2.2.1", "@ampproject/remapping@^2.1.0", "@ampproject/remapping@^2.2.0":
```
Can you please revert these changes.
The rest is looking good, I'm trying to go through #3865 to verify the output of this PR.

wgd3 commented:
With the new snapshots in place, I rebased and ran the tests for this PR again. One of the first things caught in the pipeline:

```typescript
    -     'Add Foos': props<{ foo: Foo[] }>(),
    +     'Add Foos': props<{ foos: Foo[] }>(),
    -     'Upsert Foos': props<{ foo: Foo[] }>(),
    +     'Upsert Foos': props<{ foos: Foo[] }>(),
```

This came from a change in the [actions template of the Entity schematic](https://github.com/wgd3/platform/blob/bf386a5c1322fbb460d76cecc5ccbb14b923d5e7/modules/schematics/src/entity/files/__name%40dasherize%40if-flat__/__name%40dasherize%40group-actions__.actions.ts.template#L12) that I made in one of the older commits. I reverted that with [87b34e9](https://github.com/ngrx/platform/pull/3850/commits/87b34e9a7110e9c4f01ae83587e826745ec527ce).

The other test/snapshot failure was that the order of imports had changed in the generated code for Entity:

```typescript
    -     StoreModule.forFeature(fromFoo.fooFeatureKey, fromFoo.reducer),
    -     EffectsModule.forFeature([FooEffects])
    +     EffectsModule.forFeature([FooEffects]),
    +     StoreModule.forFeature(fromFoo.fooFeatureKey, fromFoo.reducer)
```

This was remedied in [0460889](https://github.com/ngrx/platform/pull/3850/commits/0460889c28b0a1f95d7a5121cd6f441100b02874) by changing the order in which schematics are run as part of the Feature schematic.

Please let me know if you have any other feedback!

brandonroberts commented on modules/schematics/src/feature/index.spec.ts:
```diff
@@ -269,4 +271,40 @@ describe('Feature Schematic', () => {
     expect(fileContent).toMatchSnapshot();
   });
+
+  it('should create all files of a feature with an entity', async () => {
```
How about we use snapshot tests for this

timdeschryver commented on modules/schematics/src/feature/index.spec.ts:
```diff
@@ -269,4 +271,40 @@ describe('Feature Schematic', () => {
+      options,
+      appTree
+    );
+    const files = tree.files;
+    expect(
+      files.includes(`${projectPath}/src/app/foo.actions.ts`)
+    ).toBeTruthy();
+    expect(
+      files.includes(`${projectPath}/src/app/foo.actions.spec.ts`)
+    ).toBeFalsy();
+    expect(
+      files.includes(`${projectPath}/src/app/foo.reducer.ts`)
+    ).toBeTruthy();
+    expect(
+      files.includes(`${projectPath}/src/app/foo.reducer.spec.ts`)
+    ).toBeTruthy();
+    expect(
+      files.includes(`${projectPath}/src/app/foo.effects.ts`)
+    ).toBeTruthy();
+    expect(
+      files.includes(`${projectPath}/src/app/foo.effects.spec.ts`)
+    ).toBeTruthy();
+    expect(
+      files.includes(`${projectPath}/src/app/foo.selectors.ts`)
+    ).toBeFalsy();
+    expect(
+      files.includes(`${projectPath}/src/app/foo.selectors.spec.ts`)
+    ).toBeFalsy();
+    expect(files.includes(`${projectPath}/src/app/foo.model.ts`)).toBeTruthy();
```
```suggestion
    const paths = [
        `${projectPath}/src/app/foo.actions.ts`,
        `${projectPath}/src/app/foo.actions.spec.ts`,
        `${projectPath}/src/app/foo.reducer.ts`,
        `${projectPath}/src/app/foo.reducer.spec.ts`,
        `${projectPath}/src/app/foo.effects.ts`,
        `${projectPath}/src/app/foo.effects.spec.ts`,
        `${projectPath}/src/app/foo.selectors.ts`,
        `${projectPath}/src/app/foo.selectors.spec.ts`,
        `${projectPath}/src/app/foo.model.ts`,
    ];

    paths.forEach(path => {
        expect(tree.files.includes(path)).toBeTruthy();
        expect(tree.readContent(path)).toMatchSnapshot();
    })
```

timdeschryver reviewed (commented):
Sorry I forgot about this.
Thanks for the rebase, this gives the confidence that the current implementation is still working for the other configs.
As @brandonroberts mentioned, I agree that including snapshots for the generated files is helpfull.

wgd3 commented:
@timdeschryver I've removed the Actions spec file as well as the Selector/Selector spec files from the code in your commit. The Entity schematic doesn't have templates for any of these files at the moment. 

You [mentioned](https://github.com/ngrx/platform/pull/3850#discussion_r1172919625) that we should let the Entity schematic handle these things for now. If you/the team would prefer that selectors be generated as well, I can update the Feature schematic to run the Selector schematic every time.

Snapshots have been updated!

timdeschryver commented on modules/schematics/src/feature/__snapshots__/index.spec.ts.snap:
```diff
@@ -1,5 +1,177 @@
+  }),
+});
+
+export const {
+  selectIds,
+  selectEntities,
+  selectAll,
+  selectTotal,
+} = foosFeature;
```
> @timdeschryver I've removed the Actions spec file as well as the Selector/Selector spec files from the code in your commit. The Entity schematic doesn't have templates for any of these files at the moment.
> 
> You [mentioned](https://github.com/ngrx/platform/pull/3850#discussion_r1172919625) that we should let the Entity schematic handle these things for now. If you/the team would prefer that selectors be generated as well, I can update the Feature schematic to run the Selector schematic every time.
> 
> Snapshots have been updated!

Thanks! The selectors are already included within this feature file.
If I'm missing something feel free to correct me.

timdeschryver reviewed (approved):
LGTM, thanks for the updates @wgd3 !
